### PR TITLE
Add Rust support

### LINKED ISSUE #293: Add Rust to build-image
For using, you know, during builds.

Requested by Pro customer here in our helpdesk.  Others should +1 if they want this too!

- https://app.intercom.io/a/apps/q245f50x/inbox/inbox/1479299/conversations/21364638638

### PR DESCRIPTION
**Pull request by cecton**
Hello @fool 

I just started recently a new OSS project supported by Netlify: https://yewprint.rm.rs (https://github.com/cecton/yewprint) which got approved today  🎉

I'm using Rust and the build time is not short (I don't really mind personally) but I still have build time limit for the sponsored OSS account. So here I am checking if I can help getting Rust support to the Netlify image.

There was already a PR (#320) but it seems to have been closed because of lack of attention. I want to bring this back to life and add wasm-pack and wasm-bindgen support (I suppose everybody who does Rust on Netlify are here to use WASM).

Mandatory gif:

it's-alive-gif

cc @jamen @JeanBarriere @ultrasaurus 

Tasks:

 *  [x] Bring it back to life
 *  [x] Update
 *  [ ] ~Add wasm-pack and wasm-bindgen~

Closes #293 
Related to #320 
Blocked by #476

### CODE REVIEW

**Comment by cecton** [comment]
I can't even build the branch xenial locally. There is already an issue #476 

```
\e[1mdownloading: wasmer-linux-amd64
Latest release: 1.0.0-alpha3
Downloading archive from https://github.com/wasmerio/wasmer/releases/download/1.0.0-alpha3/wasmer-linux-amd64.tar.gz
######################################################################## 100.0%
\e[1minstalling: /opt/buildhome/.wasmer
Updating bash profile /opt/buildhome/.profile
we've added the following to your /opt/buildhome/.profile
If you have a different profile please add the following:
\e[2m
# Wasmer
export WASMER_DIR="/opt/buildhome/.wasmer"
[ -s "$WASMER_DIR/wasmer.sh" ] && source "$WASMER_DIR/wasmer.sh"
/opt/buildhome/.wasmer/bin/wasmer: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.27' not found (required by /opt/buildhome/.wasmer/bin/wasmer)
\e[1merror: wasmer was installed, but doesn't seem to be working :(
The command '/bin/sh -c curl https://get.wasmer.io -sSfL | sh' returned a non-zero code: 1
[1] [10:41:40] ~/r/build-image xenial > 
```

**Comment by calavera** [comment]
Hi @cecton. It might be easier for you to wait until we release https://github.com/netlify/build-image/pull/474.

We've tested it already successfully, but we're hosting a conference next week and we want to give ourselves some time to validate it widely. It should not be more than a matter of a few weeks.

**Comment by cecton** [comment]
Hello 👋 

Yes that was my intention! When the issue will be fixed I will complete the PR with wasm-pack and test on my own repo.

Also we might want to fix the version of Rust to avoid the same issue than wasmer.

Thanks a lot for the answer

**Comment by cecton** [comment]
> Also we might want to fix the version of Rust to avoid the same issue than wasmer.

We actually don't need to. rustup ensures rustc is compiled for glibc 2.11+ (64-bit Linux (kernel 2.6.32+, glibc 2.11+)). See https://doc.rust-lang.org/nightly/rustc/platform-support.html

So we can just install rust stable with rustup

File: run-build-functions.sh
Comment (mraerino) [inline run-build-functions.sh]: i think this should be specified as the build step by the user instead of being run automatically

File: run-build-functions.sh
Comment (mraerino) [inline run-build-functions.sh]: where is `repo/target` coming from?
i think in order to cache something from the user's repository you need to use `cache_cwd_directory`

File: run-build.sh
Comment (mraerino) [inline run-build.sh:41]: why are you adding this?

File: run-build-functions.sh
Comment (cecton) [reply run-build-functions.sh]: Totally agree! Should I remove the entire `if` block?

File: run-build-functions.sh
Comment (cecton) [reply run-build-functions.sh]: I don't know I just got this code from a previous PR. `target` is definitely the name of the compilation directory (a bit like `node_modules`).

So should I replace by this?

```
check_rust_target_cacheability && cache_home_directory "$(cache_cwd_directory)"/target "rust compile output"
```

File: run-build.sh
Comment (cecton) [reply run-build.sh:41]: I don't know, I'm not the author. Should I remove it?

File: run-build-functions.sh
Comment (cecton) [inline run-build-functions.sh]: This should be replaced by:

```
 mkdir -p $NETLIFY_REPO_DIR/target 
```

File: run-build-functions.sh
Comment (cecton) [reply run-build-functions.sh]: OH I see!! Nevermind what I said.

```
check_rust_target_cacheability && cache_cwd_directory target "rust compile output"
```

And I will move it up there with the others

File: run-build.sh
Comment (mraerino) [reply run-build.sh:41]: yeah, i'd remove it

File: run-build-functions.sh
Comment (mraerino) [reply run-build-functions.sh]: i think it makes sense to keep the cache handling.
if `cargo` had a command to download dependencies, but not build the project, we could execute that here, since for most languages we seperate the install and build steps.

File: run-build-functions.sh
Comment (mraerino) [reply run-build-functions.sh]: yup! great

File: run-build-functions.sh
Comment (mraerino) [reply run-build-functions.sh]: i think it can be removed entirely. `cargo` will create `target` if it doesn't exist.

File: run-build-functions.sh
Comment (cecton) [reply run-build-functions.sh]: I'm removing this so it will only restore the cache. Fixed in 60fd309

We don't pull the dependencies in Rust as they depend on the build flags.

I don't know what would be better if the dependencies were pulled separately.... let me know if something else should be done here.

File: run-build-functions.sh
Comment (cecton) [reply run-build-functions.sh]: Fixed by db4ab67

File: run-build.sh
Comment (cecton) [reply run-build.sh:41]: Fixed in e4e7354

File: run-build-functions.sh
Comment (cecton) [reply run-build-functions.sh]: Fixed in 5cd8e76

File: run-build-functions.sh
Comment (cecton) [inline run-build-functions.sh]: @mraerino Thanks a lot for your help and review btw!! :heart: 

Do you know what is the purpose of this? It seems it checks if `target` is in the `.gitignore` and if that's the case it might not restore `target` from cache.

I think it's good to have a button to purge the cache but here it seems to use a git feature (gitignore) to detect wether or not target should be cached. It doesn't sound right to me and I see no reason why you would do that.

**Comment by cecton** [comment]
@calavera @mraerino this is ready for review! :tada: 

I did not install wasm-pack by default in the end because the tooling depends on the project. Only rustup (the first main common tool to install and manage Rust toolchains) and Rust stable are installed by default.

I added a cache on rustup and `.cargo/bin` so people can change the toolchain easily (they will most likely need wasm) and their tool (wasm-pack, wasm-bindgen, ...) is cached too.

I tested with this command:

```
T=tmp/tmp.RhEPx4A6xs ./test-tools/test-build.sh ~/repos/blueprint-rs 'rustup update; cargo install wasm-pack; ./build.sh'
```

It works like a charm!

wasm-pack would probably be better in a netlify plugin so its directory can be cached also.

(Note: I had to add a line in the test script because the cache was not saved)

File: run-build-functions.sh
Comment (cecton) [inline run-build-functions.sh:60]: This is necessary because I'm caching a few subdirectories to this

File: run-build.sh
Comment (cecton) [reply run-build.sh:41]: Ah funny I have re-added this!

I know why the author added this. It's because when you run the test, the cache is restored (by `install_dependencies` in line 29) but it is not saved after the job.

File: run-build-functions.sh
Comment (cecton) [reply run-build-functions.sh]: In the end I removed it. It makes no sense.

**Review by calavera — DISMISSED** [review]
Thanks a lot for your help @cecton ! I've put this PR in one of our internal backlogs so we can make it available in an upcoming release.

**Comment by verythorough** [comment]
@cecton, thank you so much for taking the time to submit this PR and respond to reviews. As it happens, though, we've simultaneously been working on ways to reduce the size of our build image, and this PR actually increases it by about 20%.

Because build image size impacts the speed of all builds, I don't think we can justify such a significant increase for something that benefits a very small percentage of users, especially now that we have [Homebrew available on the image](https://community.netlify.com/t/now-in-experimental-alpha-homebrew-access-in-the-build-image/26074). With that in mind, I recommend you try using the Homebrew install method, and if you have issues with it, reporting them in issues here or in comments on the [Community announcement thread](https://community.netlify.com/t/now-in-experimental-alpha-homebrew-access-in-the-build-image/26074).

In the meantime, I'm going to close this PR. I'm sorry for that, as I know that communicating our priorities sooner would have saved you time and effort. I don't take that lightly!

**Comment by cecton** [comment]
hello @verythorough 

Thanks I understand. I also think it makes no sense to have such a huge container image if you want to spawn things fast.

The reason why I wanted to add it is because the build time of the image is quite high if I have to install rust and compile my entire project every time. I don't personally mind build time but Netlify counts it and it easily exceeds my OSS account. At the moment I'm have completely disabled builds.

One alternative solution would be to keep at least the caching behavior. This will take no space on your image. If we can keep rustup (a binary, I don't know the size but it should be less than 10Mb), it would be more user friendly.

I think it's worth giving a try, I don't have to do much to adapt my PR for this. Do you agree? (Unless you have other solutions to adapt the cache?)

**Comment by mraerino** [comment]
> One alternative solution would be to keep at least the caching behavior. This will take no space on your image. If we can keep rustup (a binary, I don't know the size but it should be less than 10Mb), it would be more user friendly.

i agree with this proposal. it can greatly improve how we handle Rust. moving most of the toolchain that was previously baked into the image into the cache sounds fine. i'd expect most projects to use a `rust-toolchain` file

**Comment by calavera** [comment]
Homebrew, only installs the compiler, leaving to the user to install any other tooling they need, like cargo, and clippy, which are pretty common dependencies that rustup installs by default and people need to build projects, you can look at the formula's definition:

https://formulae.brew.sh/api/formula/rust.json

I think there is still merit in some of these changes. We can follow the pattern we started with Swift and not install a version by default, Rustup will automatically install one when it detects the `rust-toolchain` file. Rustup is also a small binary that doesn't increase the image size too much.

@cecton can you make the changes to not install stable by default?

The install script should use the option `--default-toolchain none` in https://github.com/netlify/build-image/pull/477/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R493

**Comment by verythorough** [comment]
@cecton Thanks for understanding, and for your suggestion! 

Like @mraerino and @calavera I think it sounds like a worthwhile alternative that does match what we recently arrived at with Swift.

I'd also like to flag @vbrown608, @ehmicky, and @erezrokah, since they're actively working on reducing the build image size right now, and may have some other insights.

**Comment by cecton** [comment]
@verythorough @calavera @mraerino done! :tada: 

I have tested it to be sure and it is still working like a charm. :ok_hand: 

Thank you all for the time and the help. I will write a blog post at some point about how to get started with WASM, Rust and Netlify.

**Review by mraerino — DISMISSED** [review]
i think this looks great!

would you want to test this on a netlify site before we merge? i can internally set this as the build image for any site you want, you'd just have to give me a site name or id

**Comment by mraerino** [comment]
we're shortly going to have a fix for an unrelated bug that currently breaks CI on this.

**Comment by calavera** [comment]
The build should be fixed on master now. Unfortunately, you'll have to rebase this branch to get it green @cecton

**Comment by mraerino** [comment]
or you merge master in which won't require re-approval

**Comment by cecton** [comment]
> i think this looks great!
> 
> would you want to test this on a netlify site before we merge? i can internally set this as the build image for any site you want, you'd just have to give me a site name or id

Sure! Can you do it on yewprint? https://github.com/cecton/yewprint



> The build should be fixed on master now. Unfortunately, you'll have to rebase this branch to get it green @cecton

> or you merge master in which won't require re-approval

I use a merge workflow. It's fixed.


:heavy_check_mark: CI is green! cc @calavera @mraerino

**Comment by mraerino** [comment]
@cecton I updated your site to run with this PR. let us know if it works for you 🙂

**Comment by cecton** [comment]
I tested, works like a charm @mraerino 👌

**Comment by mraerino** [comment]
the team that owns the build process will give this a final look and likely release it next week. in the meantime i'll leave your site configured to use it and will switch it to the regular image once we rolled this.

**Comment by cecton** [comment]
> the team that owns the build process will give this a final look and likely release it next week. in the meantime i'll leave your site configured to use it and will switch it to the regular image once we rolled this.

Thanks! But don't worry about that because I'm stuck with manual deployment for a little longer because of some other issues.  You can change it back now if that makes your life easier @mraerino

**Comment by vbrown608** [comment]
Thanks for your contribution! The changes to the image and install script look good to me. Could you update the `Languages` section of the included software doc? https://github.com/netlify/build-image/blob/xenial/included_software.md#languages. For the version you can say `Not installed by default`. 

(I think Languages makes more sense than Included Software, since we don't list rvm, nvm, etc. there, but open to other ideas on this.)

We're planning to automate this (#441) but in the meantime I'm going to be the bot and note the image size change, which looks reasonable:
```bash
$ docker image inspect cecton-rust:latest --format='{{.Size}}'
4427251610
$ docker image inspect netlify/build:xenial --format='{{.Size}}'
4398783842
```

@rstavchansky I believe we'll also need a change to https://docs.netlify.com/configure-builds/manage-dependencies/#dependency-cache, though I don't think it needs to block merging this PR and it might even be better to run it in production as an alpha for a while before adding it to the docs site. Adding the `needs docs` label.

**Comment by cecton** [comment]
> Could you update the `Languages` section of the included software doc? https://github.com/netlify/build-image/blob/xenial/included_software.md#languages. For the version you can say `Not installed by default`.

Done


> (I think Languages makes more sense than Included Software, since we don't list rvm, nvm, etc. there, but open to other ideas on this.)

Fair enough

**Review by vbrown608 — APPROVED** [review]
Looks good :crab: ! We'll include this in the next release of the build image and make a note here when it's live. Thanks again for your work on this.

**Comment by vbrown608** [comment]
This is live in our `Xenial` build image :+1:

**Comment by mraerino** [comment]
@cecton i reset the build image on your site. let me know if there is anything amiss.

**Comment by davidbarsky** [comment]
Hi! I apologize for commenting here, but this change _appears_ to have broken our documentation previews in https://github.com/tokio-rs/tracing/pull/1130. I'm not sure what the right workaround is _or_ what the reporting mechanism y'all prefer me to use, but please let me know!

**Pull request merged**
